Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -0,0 +1,165 @@
+            if (auto* element = entry->target()) {
+                element->contentVisibilityViewportChange(entry->isIntersecting());
+                element->document().contentVisibilityDocumentState().updateOnScreenObservationTarget(*element, entry->isIntersecting());
+            }
+        }
+        return { };
+    }
+
+    ContentVisibilityIntersectionObserverCallback(Document& document)
+        : IntersectionObserverCallback(&document)
+    {
+    }
+};
+
+void ContentVisibilityDocumentState::observe(Element& element)
+{
+    auto& observer = element.document().contentVisibilityDocumentState();
+    if (auto* intersectionObserver = observer.intersectionObserver(element.document()))
+        intersectionObserver->observe(element);
+}
+
+void ContentVisibilityDocumentState::unobserve(Element& element, Document& document)
+{
+    if (auto& observer = document.contentVisibilityDocumentState().m_observer)
+        observer->unobserve(element);
+}
+
+IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Document& document)
+{
+    if (!m_observer) {
+        auto callback = ContentVisibilityIntersectionObserverCallback::create(document);
+        IntersectionObserver::Init options { &document, { }, { } };
+        auto observer = IntersectionObserver::create(document, WTFMove(callback), WTFMove(options));
+        if (observer.hasException())
+            return nullptr;
+        m_observer = observer.returnValue().ptr();
+    }
+    return m_observer.get();
+}
+
+void ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancyStatus>& relevancyToCheck)
+{
+    for (auto target : m_observer->observationTargets()) {
+        if (target) {
+            auto oldRelevancy = target->contentRelevancyStatus();
+            auto newRelevancy = oldRelevancy;
+            auto setRelevancyValue = [&](ContentRelevancyStatus reason, bool value) {
+                if (value)
+                    newRelevancy.add(reason);
+                else
+                    newRelevancy.remove(reason);
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::OnScreen))
+                setRelevancyValue(ContentRelevancyStatus::OnScreen, m_onScreenObservationTargets.contains(*target));
+
+            if (relevancyToCheck.contains(ContentRelevancyStatus::Focused))
+                setRelevancyValue(ContentRelevancyStatus::Focused, target->hasFocusWithin());
+
+            auto hasTopLayerinSubtree = [](const Element& target) {
+                for (auto& element : target.document().topLayerElements()) {
+                    if (element->isDescendantOf(target))
+                        return true;
+                }
+                return false;
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::IsInTopLayer))
+                setRelevancyValue(ContentRelevancyStatus::IsInTopLayer, hasTopLayerinSubtree(*target));
```

Comment:
Does this case work: https://github.com/WebKit/WebKit/pull/9096#discussion_r1258662305 ?

(Perhaps it does through the intersection observer, not sure)

Reply:
I created the test here:
https://github.com/web-platform-tests/wpt/pull/40970